### PR TITLE
Feat(poly): Add Dense and Sparse Multilinear Polynomials

### PR DESCRIPTION
**Pull request by PatStiles**
# Dense Multilinear Polynomial

## Description

Extends the polyomial module by adding a Dense representation of a multilinear polynomial in lagrange basis.

## Type of change

Please delete options that are not relevant.

- [x] New feature

## Checklist
- [x] Unit tests added
- [ ] Add benchmarks

### CODE REVIEW

**Review by diegokingston — APPROVED** [review]

File: math/src/polynomial/dense_multilinear_poly.rs [hunk 1]
```diff
@@ -0,0 +1,369 @@
...
+        &self.evals
+    }
+
+    #[allow(clippy::len_without_is_empty)]
+    pub fn len(&self) -> usize {
+        self.len
+    }
+
+    /// Evaluates `self` at the point `p` in O(n) time.
+    #[allow(dead_code)]
```
Comment (MauroToscano) [inline math/src/polynomial/dense_multilinear_poly.rs:54]: Allow dead code in a public function shouldn't be needed

File: math/src/polynomial/dense_multilinear_poly.rs [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+                chis[i - 1] = scalar - &chis[i];
+            }
+        }
+        if chis.len() != self.evals.len() {
+            return Err(MultilinearError::ChisAndEvalsMismatch(
+                chis.len(),
+                self.evals.len(),
+            ));
+        }
+        #[cfg(feature = "rayon")]
```
Comment (MauroToscano) [inline math/src/polynomial/dense_multilinear_poly.rs:81]: Our rayon feature is called "parallel"

File: math/src/polynomial/dense_multilinear_poly.rs [hunk 3]
```diff
@@ -0,0 +1,369 @@
...
+            return Err(MultilinearError::ChisAndEvalsMismatch(
+                chis.len(),
+                evals.len(),
+            ));
+        }
+        Ok((0..evals.len()).map(|i| &evals[i] * &chis[i]).sum())
+    }
+
+    //TODO; remove asserts
+    pub fn extend(&mut self, other: &DenseMultilinearPolynomial<F>) {
```
Comment (MauroToscano) [inline math/src/polynomial/dense_multilinear_poly.rs:114]: Add docs with ///

File: math/src/polynomial/dense_multilinear_poly.rs [hunk 4]
```diff
@@ -0,0 +1,369 @@
...
+        assert_eq!(self.evals.len(), self.len);
+        let other = other.evals.clone();
+        assert_eq!(other.len(), self.len);
+        self.evals.extend(other);
+        self.n_vars += 1;
+        self.len *= 2;
+        assert_eq!(self.evals.len(), self.len);
+    }
+
+    pub fn merge(
```
Comment (MauroToscano) [inline math/src/polynomial/dense_multilinear_poly.rs:124]: Add docs with ///

File: math/src/polynomial/dense_multilinear_poly.rs (hunk 2 above)
Comment (MauroToscano) [reply math/src/polynomial/dense_multilinear_poly.rs:81]: Nevermind, we introduced an inconsistency here, it's not a problem of this PR

File: math/src/polynomial/dense_multilinear_poly.rs (hunk 3 above)
Comment (MauroToscano) [reply math/src/polynomial/dense_multilinear_poly.rs:114]: Add panic section

**Review by MauroToscano — DISMISSED** [review]

**Review by diegokingston — DISMISSED** [review]

**Review by MauroToscano — APPROVED** [review]

File: math/src/polynomial/dense_multilinear_poly.rs [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+        let mut size = 1;
+        for j in r {
+            size *= 2;
+            for i in (0..size).rev().step_by(2) {
+                let scalar = &chis[i / 2].clone();
+                chis[i] = scalar * &j;
+                chis[i - 1] = scalar - &chis[i];
+            }
+        }
+        #[cfg(feature = "rayon")]
```
[the remark this answers is not in the document]
Comment (PatStiles) [reply math/src/polynomial/dense_multilinear_poly.rs]: resolved

**Review by MauroToscano — DISMISSED** [review]

**Review by diegokingston — APPROVED** [review]

File: math/src/polynomial/sparse_multilinear_poly.rs [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+        Ok(iter
+            .map(|i| {
+                let bits = get_bits(evals[i].0, r.len());
+                SparseMultilinearPolynomial::compute_chi(&bits, r).unwrap() * &evals[i].1
+            })
+            .sum())
+    }
+}
+
+fn get_bits(n: usize, num_bits: usize) -> Vec<bool> {
```
Comment (entropidelic) [inline math/src/polynomial/sparse_multilinear_poly.rs:108]: It would be nice to have some docs about this function

**Review by entropidelic — APPROVED** [review]

**Review by diegokingston — APPROVED** [review]

**Review by MauroToscano — APPROVED** [review]

**Pull request merged**